java/src/main/native/src/RmmJni.cpp

```diff
@@ -266,55 +274,217 @@ void set_java_device_memory_resource(
+
+ /**
+  * @brief Return pointer to the upstream resource.
+  *
+  * @return pointer to the upstream resource.
+  */
+ device_memory_resource* get_wrapped_resource() {
+  return upstream;
+ }
+
+ const device_memory_resource* get_wrapped_resource() const {
+  return upstream;
+ }
+
+ /**
+  * @brief Checks whether the upstream resource supports streams.
+  *
+  * @return true The upstream resource supports streams
+  * @return false The upstream resource does not support streams.
+  */
+ bool supports_streams() const noexcept override {
+   return upstream->supports_streams();
+ }
+
+ /**
+  * @brief Query whether the resource supports the get_mem_info API.
+  * 
+  * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+  */
+ bool supports_get_mem_info() const noexcept override { return upstream->supports_streams(); }
+
+ virtual ~logging_resource_adaptor() {
+  if (close_at_end) {
+    delete out;
+    out = nullptr;
+  }
+ }
+
+ private:
+
+ void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+   auto const p = upstream->allocate(bytes, stream);
+   auto time = std::chrono::system_clock::now();
+   std::time_t ttime = std::chrono::system_clock::to_time_t(time);
+   std::stringstream ss;
+   ss << std::put_time(std::localtime(&ttime), "%H:%M:%S")
+       << ",allocate,"
+       << p
+       << "," << bytes
+       << "," << std::to_string(reinterpret_cast<uintptr_t>(stream))
+       << std::endl;
+   {
+     std::lock_guard<std::mutex> guard(mutex);
+     *out << ss.str();
+     out->flush();
+   }
+   return p;
+ }
+
+ void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override {
+   auto time = std::chrono::system_clock::now();
+   std::time_t ttime = std::chrono::system_clock::to_time_t(time);
+   std::stringstream ss;
+   ss << std::put_time(std::localtime(&ttime), "%H:%M:%S")
+       << ",free,"
+       << p
+       << "," << bytes
+       << "," << std::to_string(reinterpret_cast<uintptr_t>(stream))
+       << std::endl;
+   {
+     std::lock_guard<std::mutex> guard(mutex);
+     *out << ss.str();
+     out->flush();
+   }
+   upstream->deallocate(p, bytes, stream);
+ }
+
+ bool do_is_equal(device_memory_resource const &other) const noexcept override {
+   if (this == &other) {
+     return true;
+   } else {
+     logging_resource_adaptor const *cast =
+        dynamic_cast<logging_resource_adaptor const *>(&other);
+     if (cast != nullptr) {
+       return upstream->is_equal(*cast->get_wrapped_resource());
+     } else {
+       return upstream->is_equal(other);
+     }
+   }
+ }
+
+ std::pair<size_t, size_t> do_get_mem_info(cudaStream_t stream) const override {
+   return upstream->get_mem_info(stream);
+ }
+
+ std::mutex mutex;
+ std::ostream * out;
+ device_memory_resource* upstream;
+ bool close_at_end;
+};
+
+// Need to keep both separate so we can shut them down appropriately
+std::unique_ptr<logging_resource_adaptor> logging_memory_resource{};
```

Comment:
Style nit: Elsewhere in this file we've been using a starting capital letter to indicate global variables and help reduce the chance of conflicting with local names.

Reply:
You are going to have to remind me every time.  That is not a convention I have ever seen used before so I always forget.